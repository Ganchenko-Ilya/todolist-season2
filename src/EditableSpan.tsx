import { ChangeEvent, KeyboardEvent, useEffect, useRef, useState } from "react";
import s from './InputAddTask.module.css'
import { TextField } from "@mui/material";
type EditableSpanPropsType = {
  title: string;
  editTitle: (newTitle: string) => void;
};
export const EditableSpan = (props: EditableSpanPropsType) => {
  const [editMode, setEditMode] = useState(false);

  const { title, editTitle } = props;

  const [value, setValue] = useState("");
  const [error, setError] = useState<string>("");

  const onDoubleClickHanlder = () => {
    setEditMode(true);
    setValue(title);
  };
  const onBlurHanlder = () => {
    if (value.trim()) {
      if (value.trim() !== title) {
        editTitle(value);
      }
      setEditMode(false);
    } else {
      setError("Incorrect input");
    }
  };
  const onChangeHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.currentTarget.value);

    setError("");
  };
  const onKeyDownHanlder = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") {
      if (value.trim()) {
        if (value.trim() !== title) {
          editTitle(value);
        }
        setEditMode(false);
      } else {
        setError("Incorrect input");
      }
    }
  };

  return (
    <div className={s.editWrapper}>
      {editMode ? (
        <TextField
          error={!!error}
          defaultValue="Hello World"
          variant="filled"
          autoFocus
          helperText={error}
          onBlur={onBlurHanlder}
          onKeyDown={onKeyDownHanlder}
          onChange={onChangeHandler}
          value={value}
          className={s.textField}
        />
      ) : (
        <span onDoubleClick={onDoubleClickHanlder}>{title}</span>
      )}
    </div>
  );
};
