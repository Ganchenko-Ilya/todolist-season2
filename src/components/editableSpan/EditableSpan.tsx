import { ChangeEvent, KeyboardEvent, useEffect, useRef, useState } from "react";
import s from "./EditableSpan.module.css";
import { TextField } from "@mui/material";
type EditableSpanPropsType = {
  title: string;
  editTitle: (newTitle: string) => void;
};
export const EditableSpan = (props: EditableSpanPropsType) => {
  const [editMode, setEditMode] = useState(false);

  const { title, editTitle } = props;

  const [value, setValue] = useState("");
  const [error, setError] = useState<string>("");

  const onDoubleClickHanlder = () => {
    setEditMode(true);
    setValue(title);
  };
  const onBlurHanlder = () => {
    if (value.trim()) {
      if (value.trim() !== title) {
        editTitle(value);
      }
      setEditMode(false);
    } else {
      setError("Incorrect input");
    }
  };
  const onChangeHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.currentTarget.value);

    setError("");
  };
  const onKeyDownHanlder = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") {
      if (value.trim()) {
        if (value.trim() !== title) {
          editTitle(value);
        }
        setEditMode(false);
      } else {
        setError("Incorrect input");
      }
    }
  };

  return (
    <div className={s.editWrapper}>
      {editMode ? (
        <TextField
          className={s.textField}
          error={!!error}
          variant="filled"
          autoFocus
          onBlur={onBlurHanlder}
          onKeyDown={onKeyDownHanlder}
          onChange={onChangeHandler}
          value={value}
          placeholder={error ? error : ""}
        />
      ) : (
        <span onDoubleClick={onDoubleClickHanlder}>{title}</span>
      )}
    </div>
  );
};
