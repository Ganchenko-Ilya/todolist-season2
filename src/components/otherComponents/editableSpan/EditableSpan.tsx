import React, { ChangeEvent, KeyboardEvent, useCallback, useEffect, useRef, useState } from "react";
import s from "./EditableSpan.module.css";
import { TextField } from "@mui/material";
type EditableSpanPropsType = {
  title: string;
  editTitle: (newTitle: string) => void;
};
export const EditableSpan = React.memo((props: EditableSpanPropsType) => {
  console.log("EditableSpan");

  const [editMode, setEditMode] = useState(false);

  const { title, editTitle } = props;

  const [value, setValue] = useState("");
  const [error, setError] = useState<string>("");

  const onDoubleClickHanlder = () => {
    setEditMode(true);
    setValue(title);
  };
  const onBlurHanlder = useCallback(() => {
    if (value.trim()) {
      if (value.trim() !== title) {
        editTitle(value);
      }
      setEditMode(false);
    } else {
      setError("Incorrect input");
    }
  }, [value, title]);
  const onChangeHandler = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.currentTarget.value);

    setError("");
  }, []);
  const onKeyDownHanlder = useCallback(
    (e: KeyboardEvent<HTMLInputElement>) => {
      if (e.key === "Enter") {
        if (value.trim()) {
          if (value.trim() !== title) {
            editTitle(value);
          }
          setEditMode(false);
        } else {
          setError("Incorrect input");
        }
      }
    },
    [value, title]
  );

  return (
    <div className={s.editWrapper}>
      {editMode ? (
        <TextField
          className={s.textField}
          error={!!error}
          variant="filled"
          autoFocus
          onBlur={onBlurHanlder}
          onKeyDown={onKeyDownHanlder}
          onChange={onChangeHandler}
          value={value}
          placeholder={error ? error : ""}
        />
      ) : (
        <span onDoubleClick={onDoubleClickHanlder}>{title}</span>
      )}
    </div>
  );
});
