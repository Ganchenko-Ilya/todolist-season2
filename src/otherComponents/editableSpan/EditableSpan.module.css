.editWrapper .textField input {
  padding: 0 5px;
  width: 120px;
  font-size: 18px;
}

.editWrapper .textField input::placeholder {
  color: red;
  font-size: 14px;
  text-align: center;
}
