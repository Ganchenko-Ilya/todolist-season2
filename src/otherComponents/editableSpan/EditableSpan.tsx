import React, { ChangeEvent, KeyboardEvent, useCallback, useState } from 'react';
import s from './EditableSpan.module.css';
import { TextField } from '@mui/material';
import { StatusType } from '../../store/app-reducer';

type EditableSpanPropsType = {
  title: string;
  statusTodo: StatusType;
  statusLoad?: StatusType;
  editTitle: (newTitle: string) => void;
};

export const EditableSpan = React.memo((props: EditableSpanPropsType) => {
  console.log('EditableSpan');
  const [editMode, setEditMode] = useState(false);
  const { title, editTitle, statusTodo, statusLoad } = props;
  const [value, setValue] = useState('');
  const [error, setError] = useState<string>('');
  const valueTrim = value.trim();

  const onDoubleClickHanlder = () => {
    setEditMode(true);
    setValue(title);
  };

  const onBlurHanlder = useCallback(() => {
    if (valueTrim) {
      if (valueTrim !== title) {
        editTitle(value);
      }
      setEditMode(false);
    } else {
      setError('Incorrect input');
    }
  }, [valueTrim, title]);

  const onChangeHandler = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.currentTarget.value);
    setError('');
  }, []);

  const onKeyDownHanlder = useCallback(
    (e: KeyboardEvent<HTMLInputElement>) => {
      if (e.key === 'Enter') {
        if (valueTrim) {
          if (valueTrim !== title) {
            editTitle(value);
          }
          setEditMode(false);
        } else {
          setError('Incorrect input');
        }
      }
    },
    [valueTrim, title]
  );

  return (
    <div className={s.editWrapper}>
      {editMode && statusTodo !== 'loading' && statusLoad !== 'loading' ? (
        <TextField
          className={s.textField}
          error={!!error}
          variant='filled'
          autoFocus
          onBlur={onBlurHanlder}
          onKeyDown={onKeyDownHanlder}
          onChange={onChangeHandler}
          value={value}
          placeholder={error ? error : ''}
        />
      ) : (
        <span onDoubleClick={onDoubleClickHanlder}>{title}</span>
      )}
    </div>
  );
});
