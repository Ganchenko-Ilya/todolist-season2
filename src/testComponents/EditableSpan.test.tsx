import renderer from "react-test-renderer";


import { EditableSpan } from "../components/otherComponents/editableSpan/EditableSpan";


test("test code Components", () => {
  const result = renderer.create(<EditableSpan editTitle={() => {}} title='text' />).toJSON();
  expect(result).toMatchSnapshot();
});