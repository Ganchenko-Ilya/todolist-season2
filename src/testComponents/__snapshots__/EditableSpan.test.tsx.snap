// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test code Components 1`] = `
<div
  className="editWrapper"
>
  <span
    onDoubleClick={[Function]}
  >
    text
  </span>
</div>
`;
